8339694: ciTypeFlow does not correctly handle unresolved constant dynamic of array type

TobiHartmann opened the pull request:
Work in progress.

Thanks,
Tobias

vnkozlov commented on src/hotspot/share/ci/ciTypeFlow.cpp:
```diff
@@ -2211,7 +2217,7 @@ bool ciTypeFlow::can_trap(ciBytecodeStream& str) {
     case Bytecodes::_ldc_w:
     case Bytecodes::_ldc2_w:
-      return str.is_in_error();
+      return str.is_in_error() || !str.get_constant().is_loaded();
```
There is also `con.is_valid()` check in `do_ldc()`. But I do know what memory is referenced in "OutOfMemoryError in the CI while loading a String constant" when it is invalid.

TobiHartmann replied:
But in that case no exception is installed and we bail out from compilation, right?
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/ci/ciTypeFlow.cpp#L746

vnkozlov replied:
Right.

iwanowww commented:
Proposed fix is broader than strictly needed to fix the immediate problem observed with condy. It affects all LDCs with not-yet-resolved CP entires. IMO it should be fine, but I haven't thought it through. 

(Also, the comment at https://github.com/openjdk/jdk/blob/master/src/hotspot/share/ci/ciTypeFlow.cpp#L2210 is outdated now.)

And [`Parse::do_one_bytecode()`](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/parse2.cpp#L1962) should always see resolved case now (`constant.is_loaded() == true`).

TobiHartmann commented:
> Proposed fix is broader than strictly needed to fix the immediate problem observed with condy

Right, to be on the safe side, I could add a `str.is_dynamic_constant()` check to limit the trap to condy. What do you think?

TobiHartmann commented:
I cleaned this up a bit more. The `iter().is_in_error()` branch in parsing is already dead without my changes because type flow will always trap. With my changes, the `!constant.is_loaded()` branch is now dead as well. I removed both and added asserts.

Given that even without my changes, we would always trap for an unresolved constant during parsing, I don't think continuing type flow analysis does make sense. In theory, it could even have a negative impact. Here's an example:

```
Class obj = ...;
if (b) {
    obj = LoadedClass.class;
} else {
    // We always trap here during parsing
    obj = UnloadedClass.class;
}
```

If we don't trap in the else branch during type flow analysis, the type of `obj` will be set to unloaded after the if. This is suboptimal because we always trap in the else branch during parsing, so the type can never be unloaded.

iwanowww reviewed (approved):
Looks good.

iwanowww commented:
> Right, to be on the safe side, I could add a str.is_dynamic_constant() check to limit the trap to condy. What do you think?

No need to. It makes sense to improve ciTypeFlow analysis for all not-yet-loaded cases. Probably, additional test cases for such cases would help, but I'd expect existing tests to provide enough coverage. So, up to you.

TobiHartmann commented:
Thanks for the review Vladimir. I think coverage of existing tests is sufficient. I did quite a few experiments and even running `java -version` with `-Xcomp -XX:-TieredCompilation` triggers such cases. It's mostly constant `.class` references to not-yet-loaded classes. I verified manually that we would already bail out during parsing in these cases even without my changes.

vnkozlov reviewed (approved):
Looks good.

TobiHartmann commented:
Thanks for the review Vladimir!

TobiHartmann commented:
/integrate
